# Week 3 - Segmenting and Clustering Neighborhoods in Toronto

Irene Bonati

### Import the necessary packages

In [1]:
import numpy as np
import pandas as pd
import requests
from bs4 import BeautifulSoup
import geocoder
from geopy.geocoders import Nominatim # Converts an address into latitude and longitude values
import folium
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe
from sklearn.cluster import KMeans
import matplotlib.cm as cm
import matplotlib.colors as colors

## Part 1

### Scrape Wikipedia website containing the list of postal codes of Canada

In [2]:
# Define URL and scrape through website
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
req = requests.get(url) 
soup = BeautifulSoup(req.content,'lxml') 

tor_neighb=[]
table=soup.find('table')
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned': # Skip if no borough is assigned
        pass
    else:
        # Define three columns: PostalCode, Borough, and Neighborhood
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        tor_neighb.append(cell) # Append name if more there is more than one neighborhood in a given borough

# Create dataframe of Toronto neighbourhoods        
df=pd.DataFrame(tor_neighb)   

# Replace name of specific cells within the dataframe
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})
df.head()

PostalCode           Borough                      Neighborhood
0        M3A        North York                         Parkwoods
1        M4A        North York                  Victoria Village
2        M5A  Downtown Toronto         Regent Park, Harbourfront
3        M6A        North York  Lawrence Manor, Lawrence Heights
4        M7A      Queen's Park     Ontario Provincial Government

In [3]:
# Print the shape of the dataframe
df.shape

(103, 3)

## Part 2

### Get the latitude and the longitude coordinates of each neighborhood and add them to the dataframe

In [4]:
def lat_long(pc):
    lati_long_coords = None
    while(lati_long_coords is None):
        g = geocoder.arcgis('{}, Toronto, Ontario'.format(pc))
        lati_long_coords = g.latlng
    return lati_long_coords

# Retrieve the postal code coordinates
postal_codes = df['PostalCode']    
coords = [lat_long(postal_code) for postal_code in postal_codes.tolist()]

# Adding columns (latitude and longitude)
df_coords = pd.DataFrame(coords, columns=['Latitude', 'Longitude'])
df['Latitude'] = df_coords['Latitude']
df['Longitude'] = df_coords['Longitude']

df.head()

PostalCode           Borough                      Neighborhood  Latitude  \
0        M3A        North York                         Parkwoods  43.75245   
1        M4A        North York                  Victoria Village  43.73057   
2        M5A  Downtown Toronto         Regent Park, Harbourfront  43.65512   
3        M6A        North York  Lawrence Manor, Lawrence Heights  43.72327   
4        M7A      Queen's Park     Ontario Provincial Government  43.66253   

   Longitude  
0  -79.32991  
1  -79.31306  
2  -79.36264  
3  -79.45042  
4  -79.39188

## Part 3

### Explore and cluster the neighborhoods in Toronto

Let's first get the geographical coordinates of Toronto:

In [5]:
address = 'Toronto, CA'

geolocator = Nominatim(user_agent="tor_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


Let's only consider the boroughs that have the word 'Toronto' in them.

In [6]:
neighb_tor= df[df['Borough'].str.contains('Toronto', na = False)].reset_index(drop=True)
neighb_tor.head()

PostalCode           Borough               Neighborhood  Latitude  Longitude
0        M5A  Downtown Toronto  Regent Park, Harbourfront  43.65512  -79.36264
1        M5B  Downtown Toronto   Garden District, Ryerson  43.65739  -79.37804
2        M5C  Downtown Toronto             St. James Town  43.65215  -79.37587
3        M4E      East Toronto                The Beaches  43.67709  -79.29547
4        M5E  Downtown Toronto                Berczy Park  43.64536  -79.37306

Let's visualize these neighborhoods in the Toronto map using Folium

In [7]:
# Create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# Add markers to map
for lat, lng, label in zip(neighb_tor['Latitude'], neighb_tor['Longitude'], neighb_tor['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Define the Foursquare credentials

In [8]:
CLIENT_ID = 'WUNQXWP4CDMN5ZXSSTACR3QNNAOVA2GLKPD2HNKOXOPPRPQ3' # your Foursquare ID
CLIENT_SECRET = 'ZEOLBKODZBJ1QWY0C55KLJVZPYPRULXAY55F033INKHM0R54' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: WUNQXWP4CDMN5ZXSSTACR3QNNAOVA2GLKPD2HNKOXOPPRPQ3
CLIENT_SECRET:ZEOLBKODZBJ1QWY0C55KLJVZPYPRULXAY55F033INKHM0R54


Let's first explore the first neighborhood(s). First, we will define its geographical coordinates.

In [9]:
print (neighb_tor.loc[0, 'Neighborhood'])

neighborhood_latitude = neighb_tor.loc[0, 'Latitude']   # neighborhood latitude value
neighborhood_longitude = neighb_tor.loc[0, 'Longitude'] # neighborhood longitude value
neighborhood_name = neighb_tor.loc[0, 'Neighborhood']   # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Regent Park, Harbourfront
Latitude and longitude values of Regent Park, Harbourfront are 43.65512000000007, -79.36263999999994.


The first neighborhoods are Regent Park and Harbourfront. Now let's look at the top 100 venues within a radius of 500 meters. To do this I created a GET request URL.

In [10]:
LIMIT = 100 # limit of number of venues returned by Foursquare API

radius = 500 # Define radius (here it is 500 m)


url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)

Define function that gets the category of a given venue.

In [11]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Let's send the GET request. The data will be collecte in a JSON file which will be then transformed into a pandas data frame.

In [12]:
results = requests.get(url).json()
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# Filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# Filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# Clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

<ipython-input-12-ea5c4b542406>:4: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  nearby_venues = json_normalize(venues) # flatten JSON


name      categories        lat        lng
0        Roselle Desserts          Bakery  43.653447 -79.362017
1           Tandem Coffee     Coffee Shop  43.653559 -79.361809
2  Figs Breakfast & Lunch  Breakfast Spot  43.655675 -79.364503
3         The Yoga Lounge     Yoga Studio  43.655515 -79.364955
4     Body Blitz Spa East             Spa  43.654735 -79.359874

Check how many venues were returned by Foursquare

In [13]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

23 venues were returned by Foursquare.


For the first neighborhood Foursquare found 23 venues. Let's define a function to get nearby venues for all the neighborhoods.

In [14]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Run the above function on each neighborhood.

In [15]:
tor_venues = getNearbyVenues(names=neighb_tor['Neighborhood'],
                                   latitudes=neighb_tor['Latitude'],
                                   longitudes=neighb_tor['Longitude']
                                  )

Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
The Danforth  East
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West
High Park, The Junction South
North Toronto West
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Rosedale
Enclave of M5E
St. James Town, Cabbagetown
First Canadi

Display venues dataframe and calculate its shape.

In [16]:
print(tor_venues.shape)
tor_venues.head()

(1703, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65512               -79.36264   
1  Regent Park, Harbourfront               43.65512               -79.36264   
2  Regent Park, Harbourfront               43.65512               -79.36264   
3  Regent Park, Harbourfront               43.65512               -79.36264   
4  Regent Park, Harbourfront               43.65512               -79.36264   

                    Venue  Venue Latitude  Venue Longitude  Venue Category  
0        Roselle Desserts       43.653447       -79.362017          Bakery  
1           Tandem Coffee       43.653559       -79.361809     Coffee Shop  
2  Figs Breakfast & Lunch       43.655675       -79.364503  Breakfast Spot  
3         The Yoga Lounge       43.655515       -79.364955     Yoga Studio  
4     Body Blitz Spa East       43.654735       -79.359874             Spa

There are a total of 1703 venues. Let's check how many venues were found for each neighborhood. This is done by grouping the venues by the different neighborhoods.

In [17]:
tor_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            63   
Brockton, Parkdale Village, Exhibition Place                           83   
CN Tower, King and Spadina, Railway Lands, Harb...                     81   
Central Bay Street                                                     64   
Christie                                                               10   
Church and Wellesley                                                   83   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             26   
Davisville North                                                       11   
Dufferin, Dovercourt Village                                           20   
Enclave of M4L                                                        100   
Enclave of M5E                                                        100   
First Canadian Place, Underground city                                100   
Forest Hill North & West                                                2   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                      49   
High Park, The Junction South                                           2   
India Bazaar, The Beaches West                                         20   
Kensington Market, Chinatown, Grange Park                              52   
Lawrence Park                                                           3   
Little Portugal, Trinity                                               41   
Moore Park, Summerhill East                                             4   
North Toronto West                                                      4   
Parkdale, Roncesvalles                                                 53   
Regent Park, Harbourfront                                              23   
Richmond, Adelaide, King                                              100   
Rosedale                                                                5   
Roselawn                                                                1   
Runnymede, Swansea                                                     39   
St. James Town                                                         78   
St. James Town, Cabbagetown                                            45   
Studio District                                                        47   
Summerhill West, Rathnelly, South Hill, Forest ...                      6   
The Annex, North Midtown, Yorkville                                    23   
The Beaches                                                             4   
The Danforth  East                                                      2   
The Danforth West, Riverdale                                            8   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         51   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             63   
Brockton, Parkdale Village, Exhibition Place                            83   
CN Tower, King and Spadina, Railway Lands, Harb...                      81   
Central Bay Street                                                      64   
Christie                                                                10   
Church and Wellesley                                                    83   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              26   
Davisville North                                                        11   
Dufferin, Dovercourt Village        

How many unique categories of venues are there?

In [18]:
print('There are {} uniques categories.'.format(len(tor_venues['Venue Category'].unique())))

There are 222 uniques categories.


Let's analyze each neighborhood.

In [19]:
# one hot encoding
tor_onehot = pd.get_dummies(tor_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
tor_onehot['Neighborhood'] = tor_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [tor_onehot.columns[-1]] + list(tor_onehot.columns[:-1])
tor_onehot = tor_onehot[fixed_columns]

tor_onehot.head()

Yoga Studio  Accessories Store  American Restaurant  Antique Shop  \
0            0                  0                    0             0   
1            0                  0                    0             0   
2            0                  0                    0             0   
3            1                  0                    0             0   
4            0                  0                    0             0   

   Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0         0            0           0                    0                 0   
1         0            0           0                    0                 0   
2         0            0           0                    0                 0   
3         0            0           0                    0                 0   
4         0            0           0                    0                 0   

   Athletics & Sports  ...  Thrift / Vintage Store  Toy / Game Store  Trail  \
0                   0  ...                       0                 0      0   
1                   0  ...                       0                 0      0   
2                   0  ...                       0                 0      0   
3                   0  ...                       0                 0      0   
4                   0  ...                       0                 0      0   

   Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0              0                              0                 0   
1              0                              0                 0   
2              0                              0                 0   
3              0                              0                 0   
4              0                              0                 0   

   Vietnamese Restaurant  Wine Bar  Wine Shop  Wings Joint  
0                      0         0          0            0  
1                      0         0          0            0  
2                      0         0          0            0  
3                      0         0          0            0  
4                      0         0          0            0  

[5 rows x 222 columns]

In [20]:
tor_onehot.shape

(1703, 222)

Let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each venue category.

In [21]:
tor_grouped = tor_onehot.groupby('Neighborhood').mean().reset_index()
tor_grouped

Neighborhood  Yoga Studio  \
0                                         Berczy Park     0.015873   
1        Brockton, Parkdale Village, Exhibition Place     0.012048   
2   CN Tower, King and Spadina, Railway Lands, Har...     0.012346   
3                                  Central Bay Street     0.000000   
4                                            Christie     0.000000   
5                                Church and Wellesley     0.012048   
6                      Commerce Court, Victoria Hotel     0.010000   
7                                          Davisville     0.000000   
8                                    Davisville North     0.000000   
9                        Dufferin, Dovercourt Village     0.000000   
10                                     Enclave of M4L     0.000000   
11                                     Enclave of M5E     0.000000   
12             First Canadian Place, Underground city     0.000000   
13                           Forest Hill North & West     0.000000   
14                           Garden District, Ryerson     0.000000   
15  Harbourfront East, Union Station, Toronto Islands     0.000000   
16                      High Park, The Junction South     0.000000   
17                     India Bazaar, The Beaches West     0.000000   
18          Kensington Market, Chinatown, Grange Park     0.000000   
19                                      Lawrence Park     0.000000   
20                           Little Portugal, Trinity     0.024390   
21                        Moore Park, Summerhill East     0.000000   
22                                 North Toronto West     0.000000   
23                             Parkdale, Roncesvalles     0.000000   
24                          Regent Park, Harbourfront     0.043478   
25                           Richmond, Adelaide, King     0.000000   
26                                           Rosedale     0.000000   
27                                           Roselawn     0.000000   
28                                 Runnymede, Swansea     0.000000   
29                                     St. James Town     0.000000   
30                        St. James Town, Cabbagetown     0.000000   
31                                    Studio District     0.000000   
32  Summerhill West, Rathnelly, South Hill, Forest...     0.000000   
33                The Annex, North Midtown, Yorkville     0.000000   
34                                        The Beaches     0.000000   
35                                 The Danforth  East     0.000000   
36                       The Danforth West, Riverdale     0.000000   
37           Toronto Dominion Centre, Design Exchange     0.000000   
38                     University of Toronto, Harbord     0.019608   

    Accessories Store  American Restaurant  Antique Shop  Aquarium  \
0            0.000000             0.000000      0.015873  0.000000   
1            0.012048             0.000000      0.000000  0.000000   
2            0.000000             0.000000      0.000000  0.000000   
3            0.000000             0.000000      0.000000  0.000000   
4            0.000000             0.000000      0.000000  0.000000   
5            0.000000             0.012048      0.000000  0.000000   
6            0.000000             0.020000      0.000000  0.000000   
7            0.000000             0.000000      0.000000  0.000000   
8            0.000000             0.000000      0.000000  0.000000   
9            0.000000             0.000000      0.000000  0.000000   
10           0.000000             0.010000      0.000000  0.000000   
11           0.000000             0.010000      0.000000  0.000000   
12           0.000000             0.020000      0.000000  0.000000   
13           0.000000             0.000000      0.000000  0.000000   
14           0.000000             0.000000      0.000000  0.000000   
15           0.000000             0.000000      0.000000  0.020408   
16           0.000000             0.000000      0.

Check the shape of the data frame (number of rows should be the same as the number of neighborhoods, 39 if only the ones with 'Toronto' are considered.

In [22]:
tor_grouped.shape

(39, 222)

For each neighborhood let's show the top 5 most common venues

In [23]:
num_top_venues = 5

for hood in tor_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = tor_grouped[tor_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0        Cocktail Bar  0.05
1         Coffee Shop  0.05
2  Seafood Restaurant  0.05
3              Bakery  0.05
4          Restaurant  0.03


----Brockton, Parkdale Village, Exhibition Place----
         venue  freq
0  Coffee Shop  0.06
1          Bar  0.06
2         Café  0.06
3   Restaurant  0.05
4    Gift Shop  0.04


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
                venue  freq
0  Italian Restaurant  0.07
1         Coffee Shop  0.06
2                Café  0.05
3   French Restaurant  0.04
4       Grocery Store  0.04


----Central Bay Street----
                       venue  freq
0                Coffee Shop  0.11
1             Clothing Store  0.06
2  Middle Eastern Restaurant  0.03
3                      Hotel  0.03
4           Sushi Restaurant  0.03


----Christie----
           venue  freq
0           Café   0.3
1  Grocery Store   0.2
2     Baby Store   

It seems like coffee shops are the most common type of venue in most of the neighborhoods explored here. Let's define a function that sorts the venues in descending order. We will use it to put the results into a dataframe.

In [24]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Let's create the new dataframe and display the top 10 venues for each neighborhood.

In [25]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = tor_grouped['Neighborhood']

for ind in np.arange(tor_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(tor_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park                Bakery   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  CN Tower, King and Spadina, Railway Lands, Har...    Italian Restaurant   
3                                 Central Bay Street           Coffee Shop   
4                                           Christie                  Café   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar           Coffee Shop    Seafood Restaurant   
1           Coffee Shop                   Bar            Restaurant   
2           Coffee Shop                  Café                   Bar   
3        Clothing Store        Cosmetics Shop                 Hotel   
4         Grocery Store           Coffee Shop    Italian Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0              Pharmacy            Restaurant           Cheese Shop   
1        Sandwich Place             Gift Shop                Lounge   
2     French Restaurant                  Park         Grocery Store   
3      Sushi Restaurant        Sandwich Place            Restaurant   
4            Baby Store            Playground           Candy Store   

       8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                        Pub        Breakfast Spot     Italian Restaurant  
1        Japanese Restaurant        Breakfast Spot              Nightclub  
2       Gym / Fitness Center             Speakeasy                 Bakery  
3  Middle Eastern Restaurant       Bubble Tea Shop                  Plaza  
4       Ethiopian Restaurant     Electronics Store            Escape Room

Let's run the k-means algorithm to cluster the neighborhoods into 4 clusters.

In [26]:
# set number of clusters
kclusters = 3

tor_grouped_clustering = tor_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(tor_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

Let's create a new dataframe that includes the cluster index and the top 10 venues for every neighborhood.

In [27]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

tor_merged = neighb_tor

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
tor_merged = tor_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

tor_merged.head() # check the last columns!

PostalCode           Borough               Neighborhood  Latitude  \
0        M5A  Downtown Toronto  Regent Park, Harbourfront  43.65512   
1        M5B  Downtown Toronto   Garden District, Ryerson  43.65739   
2        M5C  Downtown Toronto             St. James Town  43.65215   
3        M4E      East Toronto                The Beaches  43.67709   
4        M5E  Downtown Toronto                Berczy Park  43.64536   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0  -79.36264               0           Coffee Shop            Restaurant   
1  -79.37804               0           Coffee Shop        Clothing Store   
2  -79.37587               0           Coffee Shop    Italian Restaurant   
3  -79.29547               0     Health Food Store                 Trail   
4  -79.37306               0                Bakery          Cocktail Bar   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0        Breakfast Spot           Yoga Studio                Bakery   
1                  Café   Japanese Restaurant    Italian Restaurant   
2                  Café        Cosmetics Shop          Cocktail Bar   
3                   Pub            Donut Shop     Fish & Chips Shop   
4           Coffee Shop    Seafood Restaurant              Pharmacy   

       6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0                  Wine Shop    Italian Restaurant            Food Truck   
1  Middle Eastern Restaurant        Cosmetics Shop                 Hotel   
2             Lingerie Store                 Hotel            Restaurant   
3       Fast Food Restaurant        Farmers Market                  Farm   
4                 Restaurant           Cheese Shop                   Pub   

  9th Most Common Venue 10th Most Common Venue  
0           Event Space      Electronics Store  
1       Bubble Tea Shop       Ramen Restaurant  
2   Japanese Restaurant               Beer Bar  
3    Falafel Restaurant            Event Space  
4        Breakfast Spot     Italian Restaurant

Let's visualize the clusters in the map using Folium.

In [28]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(tor_merged['Latitude'], tor_merged['Longitude'], tor_merged['Neighborhood'], tor_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

Let's examine each cluster and show the venue categories that distinguish each cluster.

Cluster 1: coffee shops and food stores

In [29]:
tor_merged.loc[tor_merged['Cluster Labels'] == 0, tor_merged.columns[[1] + list(range(5, tor_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0         Downtown Toronto               0           Coffee Shop   
1         Downtown Toronto               0           Coffee Shop   
2         Downtown Toronto               0           Coffee Shop   
3             East Toronto               0     Health Food Store   
4         Downtown Toronto               0                Bakery   
5         Downtown Toronto               0           Coffee Shop   
6         Downtown Toronto               0                  Café   
7         Downtown Toronto               0                 Hotel   
8             West Toronto               0         Grocery Store   
10        Downtown Toronto               0           Coffee Shop   
11            West Toronto               0          Cocktail Bar   
12            East Toronto               0        Ice Cream Shop   
13        Downtown Toronto               0           Coffee Shop   
14            West Toronto               0                  Café   
15            East Toronto               0                  Park   
16        Downtown Toronto               0           Coffee Shop   
17            East Toronto               0           Pizza Place   
18         Central Toronto               0              Bus Line   
20         Central Toronto               0                 Hotel   
24         Central Toronto               0        Sandwich Place   
25            West Toronto               0           Coffee Shop   
26         Central Toronto               0          Dessert Shop   
27        Downtown Toronto               0                  Café   
28            West Toronto               0                  Café   
30        Downtown Toronto               0                  Café   
31         Central Toronto               0    Light Rail Station   
32        Downtown Toronto               0    Italian Restaurant   
34  Downtown Toronto Stn A               0           Coffee Shop   
35        Downtown Toronto               0           Coffee Shop   
36        Downtown Toronto               0           Coffee Shop   
37        Downtown Toronto               0           Coffee Shop   
38   East Toronto Business               0           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue        4th Most Common Venue  \
0             Restaurant        Breakfast Spot                  Yoga Studio   
1         Clothing Store                  Café          Japanese Restaurant   
2     Italian Restaurant                  Café               Cosmetics Shop   
3                  Trail                   Pub                   Donut Shop   
4           Cocktail Bar           Coffee Shop           Seafood Restaurant   
5         Clothing Store        Cosmetics Shop                        Hotel   
6          Grocery Store           Coffee Shop           Italian Restaurant   
7                   Café           Coffee Shop                          Gym   
8                   Park             Pet Store                         Bank   
10                 Hotel   Japanese Restaurant                         Park   
11                   Bar      Asian Restaurant                   Restaurant   
12          Intersection                  Park                Grocery Store   
13                  Café                 Hotel                   Restaurant   
14           Coffee Shop                   Bar                   Restaurant   
15  Fast Food Restaurant          Liquor Store               Ice Cream Shop   
16            Restaurant                  Café                        Hotel   
17               Brewery    Italian Restaurant                        Diner   
18      Business Service           Swim School                  Wings Joint   
20  Gym / Fitness Center   Japanese Restaurant                         Park   
24           Coffee Shop                  Café                Historic Site   
25                  Café      Sushi Restaurant  Eastern European Restaurant   
26        Sandwich Place            

Cluster 2: Fast food restaurants

In [30]:
tor_merged.loc[tor_merged['Cluster Labels'] == 1, tor_merged.columns[[1] + list(range(5, tor_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
19  Central Toronto               1  Fast Food Restaurant   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
19           Wings Joint            Donut Shop     Fish & Chips Shop   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
19        Farmers Market                  Farm    Falafel Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
19           Event Space  Ethiopian Restaurant            Escape Room

Cluster 3: Parks and gyms

In [31]:
tor_merged.loc[tor_merged['Cluster Labels'] == 2, tor_merged.columns[[1] + list(range(5, tor_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
9   East York/East Toronto               2                  Park   
21         Central Toronto               2                  Park   
22            West Toronto               2                  Park   
23         Central Toronto               2              Gym Pool   
29         Central Toronto               2                   Gym   
33        Downtown Toronto               2                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
9           Intersection           Wings Joint            Donut Shop   
21      Business Service           Wings Joint   Dumpling Restaurant   
22        Sandwich Place           Wings Joint            Donut Shop   
23    Photography Studio                  Park            Playground   
29                  Park            Playground          Tennis Court   
33   Japanese Restaurant            Playground            Bike Trail   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
9      Fish & Chips Shop  Fast Food Restaurant        Farmers Market   
21     Fish & Chips Shop  Fast Food Restaurant        Farmers Market   
22  Fast Food Restaurant        Farmers Market                  Farm   
23               Dog Run        Farmers Market                  Farm   
29           Wings Joint               Dog Run        Farmers Market   
33            Donut Shop  Fast Food Restaurant        Farmers Market   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
9                   Farm    Falafel Restaurant            Event Space  
21                  Farm    Falafel Restaurant            Event Space  
22    Falafel Restaurant           Event Space   Ethiopian Restaurant  
23    Falafel Restaurant           Event Space   Ethiopian Restaurant  
29                  Farm    Falafel Restaurant            Event Space  
33                  Farm    Falafel Restaurant            Event Space